fpdf/text_region.py

```diff
@@ -0,0 +1,428 @@
+        print_sh: bool = False,
+        skip_leading_spaces: bool = False,
+        wrapmode: WrapMode = None,
+        **kwargs,
+    ):
+        self.pdf = pdf
+        self.align = Align.coerce(align)  # default for auto paragraphs
+        self.line_height = line_height
+        self.print_sh = print_sh
+        self.wrapmode = WrapMode.coerce(wrapmode)
+        self.skip_leading_spaces = skip_leading_spaces
+        self._paragraphs = []
+        self._active_paragraph = None
+        super().__init__(pdf, *args, **kwargs)
+        if text:
+            self.write(text)
+
+    def __enter__(self):
+        if self.pdf.is_current_text_region(self):
+            raise FPDFException(
+                f"Unable to enter the same {self.__class__.__name__} context recursively."
+            )
+        self._page = self.pdf.page
+        self.pdf._push_local_stack()
+        self.pdf.page = 0
+        self.pdf.register_text_region(self)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.pdf.clear_text_region()
+        self.pdf.page = self._page
+        self.pdf._pop_local_stack()
+        self.render()
+
+    def _check_paragraph(self):
+        if self._active_paragraph == "EXPLICIT":
+            raise FPDFException(
+                "Conflicts with active paragraph. Either close the current paragraph or write your text inside it."
+            )
+        if self._active_paragraph is None:
+            p = Paragraph(
+                region=self,
+                align=self.align,
+                skip_leading_spaces=self.skip_leading_spaces,
+            )
+            self._paragraphs.append(p)
+            self._active_paragraph = "AUTO"
+
+    def write(self, text: str, link=None):  # pylint: disable=unused-argument
+        self._check_paragraph()
+        self._paragraphs[-1].write(text)
+
+    def ln(self, h=None):
+        self._check_paragraph()
+        self._paragraphs[-1].ln(h)
+
+    def paragraph(
+        self,
+        align=None,
+        line_height=None,
+        skip_leading_spaces: bool = False,
+        top_margin=0,
+        bottom_margin=0,
+        wrapmode: WrapMode = None,
+    ):
+        if self._active_paragraph == "EXPLICIT":
+            raise FPDFException("Unable to nest paragraphs.")
+        p = Paragraph(
+            region=self,
+            align=align or self.align,
+            line_height=line_height,
+            skip_leading_spaces=skip_leading_spaces or self.skip_leading_spaces,
+            wrapmode=wrapmode,
+            top_margin=top_margin,
+            bottom_margin=bottom_margin,
+        )
+        self._paragraphs.append(p)
+        self._active_paragraph = "EXPLICIT"
+        return p
+
+    def end_paragraph(self):
+        if not self._active_paragraph:
+            raise FPDFException("No active paragraph to end.")
+        # self._paragraphs[-1].write("\n")
+        self._active_paragraph = None
+
+
+class TextRegion(ParagraphCollectorMixin):
```

Comment:
Maybe this class should inherit from `abc.ABC` to indicate that it's abstract?
https://docs.python.org/3/library/abc.html

Reply:
It's been a while since I last studied the use of ABCs...

I'm not confident enough about it right now. Another topic to revisit later.

Reply:
Alright